### ISSUE TITLE
Focus Areas for Q4 2023

### ISSUE THREAD

**Issue by abdallahshaban557** [aws-amplify/amplify-js#11833]
Hello everyone,

We wanted to inform you we have published a page in our wiki detailing our [Focus Areas for 2023](https://github.com/aws-amplify/amplify-js/wiki/Focus-Areas-Q4-2023). We wanted to make sure you know what we are working on, and get your feedback as we move forward.

**Comment by rpostulart** [comment]
https://github.com/aws-amplify/amplify-swift/issues/1005

Is this still on the backlog?

**Comment by codingnuclei** [comment]
Really interested in the webAuthN solution. We have managed to build a solution with cognito and amplifyJS that enables user to use a one time code or their biometrics.  

The tricky part with the webauthn was the server aspect of  how cognito lambda triggers don't pass session data been the triggers you need it to.

Then there is a tricky Ui/Ux challenge. Unless it changed recently you can not tell if a user has a passkey for your domain.  Meaning how you present the auth flow to the user can be challenging.  Conditional UI does aid slightly but still requires a user to add there username.
>
Interested to see your approach for both aspects👍

**Comment by abdallahshaban557** [comment]
Hi @codingnuclei  - thank you so much for this feedback - this is coming at an excellent time since we are going through our design. We are excited to share more as we have our design in a better spot!

**Comment by codingnuclei** [comment]
@abdallahshaban557 excellent. Would be happy to set some time aside to discuss our findings further. Just reach out if needed!

**Comment by abdallahshaban557** [comment]
Got it @codingnuclei - will reach out when we are at a closer point to sharing our designs!

**Comment by oe-bayram** [comment]
We use the Amplify Framework for creating large ERP systems and encounter some scalability limitations, especially concerning these issues:
- aws-amplify/amplify-category-api/issues/32
- aws-amplify/amplify-cli/issues/6898

What's the status of these issues? Will they also be included in the Focus Areas for Q4 2023?

**Comment by abdallahshaban557** [comment]
Hi @oe-bayram  - Thank you for reaching out!

on the stack mapping issue, our team is still investigating this with the support of the AppSync team to overcome this limitation. We will keep providing updates on that issue when we have a path forward. We do not have a clear ETA yet.

For the Plugins support - can you please elaborate on your use case here? What are you looking for specifically?

**Comment by abdallahshaban557** [comment]
@SilverLinings89 - we really appreciate you taking the time to give us this feedback. I will address them individually:

- We are exploring what an out of the box solution for multi-tenancy is. We do not yet have a solid solution, but we will have more to share once a design is in place to get feedback from our community. While this did not make it for 2023, it is top of mind for us in 2024.
- For account merging, we worked with the Amazon Cognito team, and they have recently enabled support for that in the preSignUp lambda trigger. So you can [take a look at this approach](https://github.com/aws-amplify/amplify-js/issues/3431#issuecomment-1505507067) we've provided developers wanting to use this feature. We are considering providing an "out of the box" experience, but are still debating how to do it in the most safe way possible.
- We're optimzing tree-shaking and reducing bundle size substantially in our next major version. Can you elaborate a bit more on the warnings you are getting?
- The React Native Authenticator supporting social sign in is happening very soon! It will be released by the time we get our new major version of the Amplify JavaScript out. The Amplify UI team has been working with us to enable that as fast as they could.
- The type decleration issue will be fixed as part of our overall "TypeScript Enhancement" initiative. We will also provide a new client that can be used in a fully typesafe way, which is generated from your schema definition! 
- On the issues with regressions, we had discovered some of our internal processes needed tightening to make sure we release code safely, and that all breaking changes/regressions are caught and avoided. We are getting better at preventing these issues everyday!

**Comment by abdallahshaban557** [comment]
That is amazing to hear - we hope to be bringing even more value to you and your team soon!

Got it on the issue, and it is on our radar! We already have a [Github issue](https://github.com/aws-amplify/amplify-js/issues/6202) on this as well! @AllanZhengYP on our team is actively investigating it!. We will provide updates on that Github issue as we make progress.

**Comment by DougalW** [comment]
Please please please prioritise the CloudFormation scaling issues:

https://github.com/aws-amplify/amplify-category-api/issues/32
https://github.com/aws-amplify/amplify-category-api/issues/658
https://github.com/aws-amplify/amplify-studio/issues/414
https://github.com/aws-amplify/amplify-cli/issues/13207

There's clearly something going on under the covers that's preventing scaling past a certain point. It seems to start around 40 tables in dynamoDB and become unusable somewhere >50, so much so that Amplify becomes unusable.

**Comment by abdallahshaban557** [comment]
Hi @DougalW  - the issues you have mentioned are on our radar, we are aware of these scaling issues, and have some potential paths forward identified to help us alleviate these pain points. We will continue to provide updates on these Github issues as we make more progress.

**Comment by aaamoshd** [comment]
Hello there. It would be great, if the team could also focus on changing the behavior of Device Tracking when Custom_Auth flow is used. As it stands, device tracking is a feature that cannot be used in tandem with Custom_Auth flow which might be very disruptive for certain implementations.

Here are the related cases with Device Tracking and Custom_Auth flow. 

https://github.com/aws-amplify/amplify-js/issues/9592
https://github.com/aws-amplify/amplify-js/issues/12220

**Comment by abdallahshaban557** [comment]
Hi @aaamoshd  - thank you so much for this feedback. We had been investigating them before - and agree that not having them supported is disruptive. We will continue to provide updates on them as we make progress.

**Comment by jeffpball** [comment]
Hello, lots of love to the AWS Amplify team and AWS more widely. I know you're a business, but your tools enable developers to do a lot of cool stuff. So thanks for your work.

This is probably a lot lower prio than the folks who are working on production code bases using Next.js, but I wanted to bring it up on behalf of Next.js novices like myself. 

It would be great to have [this Next.js tutorial](https://docs.amplify.aws/start/getting-started/installation/q/integration/next/) in the Amplify docs be updated to include:

- AWS Amplify JavaScript v6 Developer Preview usage
- App Router (maybe alongside Pages Router like you've done elsewhere in Amplify docs)
- TypeScript examples alongside JavaScript

It seems to me that specifically [this section](https://docs.amplify.aws/start/getting-started/data-model/q/integration/next/#api-with-server-side-rendering-ssr) could use the updates mentioned above. There's probably other places in the tutorial that will need updating, but if we could get a cutting edge Next.js GraphQL CRUD app (including Authentication) server component tutorial with all the recent updates applied, it would be a big win for me personally and hopefully for others.

A little backstory: I first attempted to ask Vercel to update their Amplify example app repo and they decided to remove it and refer people to your docs. See here https://github.com/vercel/next.js/issues/57127 and here https://github.com/vercel/next.js/commit/ac54377e6723543333109adff43a804778712a8f

So, now I'm trying to piece together both Vercel's primary documentation regarding App Router and these various Amplify docs:
[Amplify JS 6 Developer Preview announcement](https://aws.amazon.com/blogs/mobile/aws-amplify-javascripts-new-developer-preview-with-reduced-bundle-sizes-improved-typescript-and-next-js-support/)
[The actual JS 6 Developer Preview docs](https://docs.amplify.aws/lib-v1/project-setup/create-application/q/platform/js/)
[Next.js Features blog article](https://aws.amazon.com/blogs/mobile/5-next-js-features-that-are-better-with-aws-amplify/)

...and a few other docs that are out there. Thanks in advance for any assistance on this!

**Comment by abdallahshaban557** [comment]
Hi @jeffpball - we are working on updating our documentation to make it a lot easier to use NextJS with Amplify as part of our v6 release of Amplify JavaScript. We understand that with the release of App Router the paradigm has shifted a bit on how to use NextJS, and our tooling is now catching up to make that experience easier. Stay tuned! V6 and our updated docs are almost there!

**Comment by jeffpball** [comment]
> Hi @jeffpball - we are working on updating our documentation to make it a lot easier to use NextJS with Amplify as part of our v6 release of Amplify JavaScript. We understand that with the release of App Router the paradigm has shifted a bit on how to use NextJS, and our tooling is now catching up to make that experience easier. Stay tuned! V6 and our updated docs are almost there!

That's great news, thanks a lot @abdallahshaban557 !

**Comment by rogueturnip** [comment]
Excited to hear it's "almost there" :)

**Comment by charlieforward9** [comment]
I still lack many of the 'behind the scenes' insights, but it seems that Amplify's interoperability with Web Workers is not stable. Having a bug-free experience when using Amplify API's and Web Workers is necesarry for our use case. More information on this issue: aws-amplify/amplify-js#12278

**Comment by RichiCoder1** [comment]
Wanted to +1 how excited I'd personally before Passkey/WebAuthN support being first class. Massive enabler for ditching passwords without having to rely on third party social providers.

**Comment by djorgji** [comment]
+1 for the passwordless logins for us.

**Comment by osehmathias** [comment]
Is Datastore encryption on the client on the horizon? Or should we be generously tempering our expectations in this regard?

**Comment by cwomack** [comment]
Hello everyone! We appreciate all the comments and feedback on this issue thus far, and wanted to give a quick status update/summary of what was accomplished in Q4 of 2023 per these focus areas.  We had an exciting quarter to close out last year and saw “wins” in the following areas:

- [v6 General Availability](https://aws.amazon.com/blogs/mobile/amplify-javascript-v6/) - TypeScript Strict mode and no implicit any’s

- [Bundle size improvements](https://github.com/aws-amplify/amplify-js/issues?q=label%3Abundle-size+is%3Aclosed)

- [Next Gen DevEx (Gen 2)](https://aws.amazon.com/blogs/mobile/introducing-amplify-gen2/) - Amplify’s Code First Developer experience (now in developer preview)

- [Support for Hosting any SSR App](https://aws.amazon.com/blogs/mobile/introducing-support-for-hosting-any-ssr-app-on-aws-amplify-hosting/)

- [Wildcard Subdomains](https://aws.amazon.com/blogs/mobile/wildcard-subdomains-for-multi-tenant-apps-on-aws-amplify-hosting/)

- [Creating GraphQL API’s for any existing MySQL and PostgreSQL Databases](https://aws.amazon.com/blogs/mobile/create-a-graphql-api-for-any-existing-mysql-and-postgresql-database/)

We also wanted to recognize the areas where we didn’t make quite as much progress as we would have liked, such as the Storage category improvements, observability/monitoring, and enabling MagicLink/one-time password.  We plan on having these be a part of the Q1 2024 Focus Areas and will release a new issue that details these by the end of this month!
